src/bit_vec.rs

```diff
@@ -50,21 +64,48 @@ impl<O: BitOrder, T: BitStore + ToByteSlice> Encode for BitVec<O, T> {
 			"Attempted to serialize a collection with too many elements.",
 		);
 		Compact(len as u32).encode_to(dest);
-		dest.write(self.as_slice().as_byte_slice());
+
+		let byte_slice: &[u8] = self.as_slice().as_byte_slice();
+
+		if cfg!(target_endian = "big") && mem::size_of::<T>() != 1 {
+			let mut vec_u8: Vec<u8> = byte_slice.into();
+			reverse_endian(&mut vec_u8, mem::size_of::<T>());
+			dest.write(&vec_u8);
+		} else {
+			dest.write(byte_slice);
+		}
 	}
 }
 
 impl<O: BitOrder, T: BitStore + ToByteSlice> EncodeLike for BitVec<O, T> {}
 
+/// # WARNING
+///
+/// In bitvec v0.17.4 the only implementation of BitStore are u8, u16, u32, u64, usize.
+/// This implementation actually only support u8, u16, u32 and u64, encoding of with BitStore being
+/// usize is inconsistent between plateform.
 impl<O: BitOrder, T: BitStore + FromByteSlice> Decode for BitVec<O, T> {
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
 		<Compact<u32>>::decode(input).and_then(move |Compact(bits)| {
 			let bits = bits as usize;
 			let required_bytes = required_bytes::<T>(bits);
 
-			let vec = read_vec_u8(input, required_bytes)?;
+			let mut vec_u8 = read_vec_from_u8s::<I, u8>(input, required_bytes)?;
+
+			if cfg!(target_endian = "big") && mem::size_of::<T>() != 1 {
+				reverse_endian(&mut vec_u8, mem::size_of::<T>());
+			}
+
+			let mut aligned_vec: Vec<T> = vec![0u8.into(); required_bytes / mem::size_of::<T>()];
 
-			let mut result = Self::from_slice(T::from_byte_slice(&vec)?);
+			unsafe {
+				let aligned_u8_ptr = aligned_vec.as_mut_ptr() as *mut u8;
+				for (i, v) in vec_u8.iter().enumerate() {
+					*aligned_u8_ptr.add(i) = *v;
+				}
+			}
```

Comment:
Why don't we use the output of `read_vec_from_u8s` directly as input for `from_vec`?

Reply:
so above we called `read_vec_from_u8s::<I, u8>`, we can't directly do `read_vec_from_u8s::<I, T>` (where T is `BitStore` thus one of u8, u16, ..., u64) because T doesn't implement `ToMutByteSlice`.

(it implements`Clone`, and we could get around the `Default` requirement in `read_vec_from_u8s`, so `ToMutByteSlice` is difficult without changing the signature `impl<O: BitOrder, T: BitStore + ToByteSlice> Encode for BitVec<O, T>`).

So as a result of having used `u8` we might not have an aligned vec (at least under miri test condition). so here I just copy it to an aligned vec.

But anyway the mirit test is failing because of outdated bitvec. But I thought fixing the alignment would be better even if it involve an additional copy here.
Note that once we bump bitvec we can factorize it quite easily, removing this unsafe and decode a `Vec<T>` directly.

Reply:
Makes sense! Ty for explaining it!